dagutils/diff.go

```diff
@@ -100,69 +103,183 @@ func ApplyChange(ctx context.Context, ds ipld.DAGService, nd *dag.ProtoNode, cs
 // It only traverses links in the following cases:
 // 1. two node's links number are greater than 0.
 // 2. both of two nodes are ProtoNode.
+// 3. there is only one link under the name being traversed.
 // Otherwise, it compares the cid and emits a Mod change object.
-func Diff(ctx context.Context, ds ipld.DAGService, a, b ipld.Node) ([]*Change, error) {
+func Diff(ctx context.Context, ds ipld.DAGService, a, b *dag.ProtoNode) ([]*Change, error) {
+	// Short-circuit the identity case to avoid inspecting links.
 	if a.Cid() == b.Cid() {
 		return []*Change{}, nil
 	}
 
-	cleanA, okA := a.Copy().(*dag.ProtoNode)
-	cleanB, okB := b.Copy().(*dag.ProtoNode)
-
-	linksA := a.Links()
-	linksB := b.Links()
-
-	if !okA || !okB || (len(linksA) == 0 && len(linksB) == 0) {
-		return []*Change{{Type: Mod, Before: a.Cid(), After: b.Cid()}}, nil
+	var out []*Change
+	if !bytes.Equal(a.Data(), b.Data()) {
+		// FIXME: This is a very dirty way of reporting a difference in Data()
+		//  but the current `Change` API doesn't support anything but links.
+		out = append(out, &Change{Type: Mod, Path: "<NODE-DATA>", Before: a.Cid(), After: b.Cid()})
```

Comment:
Wrote some thoughts and relevant examples in https://github.com/ipfs/go-ipfs/issues/4801#issuecomment-1024303462

Perhaps switching to Logical Format described there, with `Data` and `Links/<index>/Name` prefixes in `Change.Path` would remove the need for hacks like this.

Having that, `ipfs object diff` would ignore this additional data,
but `ipfs dag diff` could take advantage of it and show more detailed DAG diff.

Reply:
@lidel I'm not sure I follow the concrete direction you're requesting there; I think I'll need that to be lowered a bit more to the codebase I'm dealing with here.

From what I'm understanding in the linked proposal a `dag diff` would now be:
* `dag get` to the JSON format already in place
* diff the JSON outputs

Which would operate completely in the `go-ipfs` repo and not here (this work would could apply as the new, a little less broken, `object diff` command if you want).

Reply:
Indeed, generating the `dag diff` based on `dag-json` representation would make things easier and less error-prone. 
Since everything would be serialized to `dag-json` representation, we could support `dag-pb`, `raw` amd `dag-cbor` CIDs out of the box. 

:point_right:  @schomatis  I think we should  give that a try and park this PR for now – focus on go-ipfs and its JSON-Patch-based `dag diff` instead.

Regarding this PR (when we revisit it later)
- is `Diff` only used in `object diff` or are there other places where we use diff implementation from here?
- if unused, perhaps we should remove `Diff` and `DiffEnumerate`, and switch `ipfs object diff` to use `dag diff` behind the scenes?